swarm/src/lib.rs

```diff
@@ -1731,3 +1777,134 @@ mod tests {
         }))
     }
 }
+
+pub mod dial_opts {
+    use libp2p_core::{Multiaddr, PeerId};
+
+    #[derive(Debug)]
+    pub enum DialOpts {
+        // TODO:  How about folding these structs into the enum variants?
+        WithPeerId(WithPeerId),
+        WithPeerIdWithAddresses(WithPeerIdWithAddresses),
+        WithoutPeerIdWithAddress(WithoutPeerIdWithAddress),
+    }
+
+    impl DialOpts {
+        pub fn peer_id(peer_id: PeerId) -> WithPeerId {
+            WithPeerId {
+                peer_id,
+                condition: Default::default(),
+            }
+        }
+
+        pub fn unknown_peer_id() -> WithoutPeerId {
+            WithoutPeerId {}
+        }
+
+        pub fn get_peer_id(&self) -> Option<PeerId> {
+            match self {
+                DialOpts::WithPeerId(WithPeerId { peer_id, .. }) => Some(peer_id),
+                DialOpts::WithPeerIdWithAddresses(WithPeerIdWithAddresses { peer_id, .. }) => {
+                    Some(peer_id)
+                }
+                DialOpts::WithoutPeerIdWithAddress(_) => None,
+            }
+        }
+    }
+
+    #[derive(Debug)]
+    pub struct WithPeerId {
+        pub(crate) peer_id: PeerId,
+        pub(crate) condition: PeerCondition,
+    }
+
+    impl WithPeerId {
+        pub fn condition(mut self, condition: PeerCondition) -> Self {
+            self.condition = condition;
+            self
+        }
+
+        pub fn addresses(self, addresses: Vec<Multiaddr>) -> WithPeerIdWithAddresses {
+            WithPeerIdWithAddresses {
+                peer_id: self.peer_id,
+                condition: self.condition,
+                addresses,
+                extend_addresses_through_behaviour: false,
+            }
+        }
+
+        pub fn build(self) -> DialOpts {
+            DialOpts::WithPeerId(self)
+        }
+    }
+
+    #[derive(Debug)]
+    pub struct WithPeerIdWithAddresses {
+        pub(crate) peer_id: PeerId,
+        pub(crate) condition: PeerCondition,
+        pub(crate) addresses: Vec<Multiaddr>,
+        pub(crate) extend_addresses_through_behaviour: bool,
+    }
+
+    impl WithPeerIdWithAddresses {
```

Comment:
```suggestion
    impl WithPeerIdWithAddresses {
        pub fn condition(mut self, condition: PeerCondition) -> Self {
            self.condition = condition;
            self
        }
        
```
Maybe add this method here as well.

Reply:
I am a bit reluctant to offer two ways to achieve the same thing. In what use-case would one want to call `condition` on `WithPeerIdWithAddresses` and couldn't do so on `WithPeerId`?

Reply:
I am also a bit undecided in this aspect. 
But the `DialOpts` are trying to follow the syntax of a builder pattern, and with a builder it normally doesn't matter in which order you specify the paramters, so you would expect that `DialOpts::peer_id(...).condition(..).addresses(..)` and `DialOpts::peer_id(...).addresses(..).condition(..)` both work. 
That being said, the current implementation doesn't strictly follow the builder pattern anyway (at least not the way it is described in the [Rust API Guidelines](https://rust-lang.github.io/api-guidelines/type-safety.html#builders-enable-construction-of-complex-values-c-builder)), therefore I also see the point of avoiding duplicated code.

Reply:
Addressed in https://github.com/libp2p/rust-libp2p/pull/2317/commits/26dedbec870842a5b3f9299894db14790b759777.